### PR TITLE
Suggest debugger behavior when icount stepping over restarting instruction

### PR DESCRIPTION
**Pull request by timsifive**
Tons of mailing list discussion, e.g. https://lists.riscv.org/g/tech-debug/message/1016

### CODE REVIEW

File: debugger_implementation.tex [hunk 1]
```diff
@@ -615,3 +615,32 @@ \subsection{Single Step} \label{nativestep}
...
+
+To help users out, debuggers should detect when a single step restarted an
+instruction, and then step again. This way the users see the expected behavior
+of stepping over the instruction. Ideally the debugger would notify the user
+that an exception handler executed the first time.
+
+\begin{steps}{The debugger should perform this extra step if all the following
+        conditions are satisfied:}
+    \item The PC did not change during the first step.
+    \item The PC does not equal the {\tt epc} register for the current privilege mode.
```
Comment (pdonahue-ventana) [inline debugger_implementation.tex:636]: This condition isn't immediately obvious to me.  If you're stepping S mode, PC=X, you step, and PC is still X then step one more time.  Why do we also need to check if sepc=X?

Comment (timsifive) [reply debugger_implementation.tex:636]: That should be `tvec`, not `epc`.

The condition I'm worried about is if you're stepping the first instruction in an exception handler, that itself causes that exception handler to be called. But maybe that doesn't matter similar to jal to self makes no changes after the first invocation.

I'd love some more input into my thinking here.

Comment (pdonahue-ventana) [reply debugger_implementation.tex:636]: tvec makes more sense.  Say that you're stepping U+S modes and breakpoints go to M.  The following happens:
- You step X and it traps.
- The trap sends you to M mode with mepc=X.
- PC did change so you wait for the debugger user to request another step.  (Don't automatically do the extra step you're describing.)
- User says to step.  Your vector is an illegal instruction (or whatever) so it traps.
- The trap sends you to M mode with mepc=stvec.

You're saying that the debugger should not do the extra step here.  I agree that you don't want to do the extra step.  However, if you did then you'd hit the same illegal instruction and come back to M mode with mepc=stvec so there's no real harm.

**Comment by pdonahue-ventana** [comment]
Modulo the tvec details, I think that this is generally good.  If you get a page fault on a load, the debugger should step twice because the PC didn't advance.  If you get an interrupt on a nop, the debugger should also step twice because the PC didn't advance.  If you get an illegal instruction on something that's emulated, the debugger will naturally only step once because the PC did advance.  No handler needs to do any fixup and all the work is pushed to the debugger which presumably isn't as performance-critical as one of the handlers.

**Comment by timsifive** [comment]
I think we might not need the `tvec` rule at all. Then all the debugger needs to check is whether the PC changed. This works, iff the following statement is true: Every RISC-V instruction changes the PC or has no side effects when repeated.

Is that a true statement?

**Comment by pdonahue-ventana** [comment]
This isn't just caused by exceptions like page faults.  Interrupts also can cause this behavior on otherwise innocuous instructions.  The solution is the same, but I would mention both exceptions and interrupts in the description just so everyone is aware of the full spectrum of cases.

**Comment by timsifive** [comment]
I've added interrupts as another source of this problem.

**Review by pdonahue-ventana — APPROVED** [review]
Looks good.

**Pull request merged**